tests/test_class_sh_void_ptr_capsule.cpp

```diff
@@ -7,122 +7,96 @@
 namespace pybind11_tests {
 namespace class_sh_void_ptr_capsule {
 
-// Conveniently, the helper serves to keep track of `capsule_generated`.
-struct HelperBase {
-    HelperBase() = default;
-    HelperBase(const HelperBase &) = delete;
-    virtual ~HelperBase() = default;
-
-    bool capsule_generated = false;
-    virtual int get() const { return 100; }
-};
+struct Valid {};
 
-struct Valid : public HelperBase {
-    int get() const override { return 101; }
+struct NoConversion {};
 
-    PyObject *as_pybind11_tests_class_sh_void_ptr_capsule_Valid() {
-        void *vptr = dynamic_cast<void *>(this);
-        capsule_generated = true;
-        // We assume vptr out lives the capsule, so we use nullptr for the
-        // destructor.
-        return PyCapsule_New(vptr, "::pybind11_tests::class_sh_void_ptr_capsule::Valid", nullptr);
-    }
-};
+struct NoCapsuleReturned {};
 
-struct NoConversion : public HelperBase {
-    int get() const override { return 102; }
-};
+struct AsAnotherObject {};
 
-struct NoCapsuleReturned : public HelperBase {
-    int get() const override { return 103; }
+// Create a void pointer capsule for test. The encapsulated object does not
+// matter for this test case.
+PyObject *create_test_void_ptr_capsule() { return PyCapsule_New((void *) 12345, "int", nullptr); }
```

Comment:
While this is evidently fine on all platforms today, I'm afraid casting an arbitrary integer to a pointer could trigger some sanitizer error in the future. Ideas:
```
// Deemed safe not to incref. None surely outlives this capsule.
static_cast<void *>(Py_None)
```
or
```
static int just_to_have_something_to_point_to = 0;
static_cast<void *>(&just_to_have_something_to_point_to)
```